samples/subsys/ipc/openamp/src/main.c

```diff
@@ -0,0 +1,294 @@
+{
+	(void)rpmsg_create_ept(ep, rdev, name,
+			RPMSG_ADDR_ANY, dest,
+			endpoint_cb,
+			rpmsg_service_unbind);
+
+	k_sem_give(&ept_sem);
+}
+
+static unsigned int receive_message(void)
+{
+	while (k_sem_take(&data_rx_sem, K_NO_WAIT) != 0) {
+		int status = k_sem_take(&data_sem, K_FOREVER);
+
+		if (status == 0) {
+			virtqueue_notification(vq[0]);
+		}
+	}
+	return received_data;
+}
+
+static int send_message(unsigned int message)
+{
+	return rpmsg_send(ep, &message, sizeof(message));
+}
+
+static struct rpmsg_virtio_shm_pool shpool;
+
+void app_task(void *arg1, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	int status = 0;
+	unsigned int message = 0;
+	struct metal_device *device;
+	struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+
+	printk("\r\nOpenAMP[master] demo started\r\n");
+
+	metal_init(&metal_params);
+	if (status != 0) {
+		printk("metal_init: failed - error code %d\n", status);
+		return;
+	}
+
+	status = metal_register_generic_device(&shm_device);
+	if (status != 0) {
+		printk("Couldn't register shared memory device: %d\n", status);
+		return;
+	}
+
+	status = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (status != 0) {
+		printk("metal_device_open failed: %d\n", status);
+		return;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (io == NULL) {
+		printk("metal_device_io_region failed to get region\n");
+		return;
+	}
+
+	/* setup IPM */
+	ipm_handle = device_get_binding("MAILBOX_0");
+	if (ipm_handle == NULL) {
+		printk("device_get_binding failed to find device\n");
+		return;
+	}
+
+	ipm_register_callback(ipm_handle, platform_ipm_callback, NULL);
+
+	ipm_set_enabled(ipm_handle, 1);
+	if (status != 0) {
+		printk("ipm_set_enabled failed\n");
+		return;
+	}
+
+	/* setup vdev */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
+	if (vq[0] == NULL) {
+		printk("virtqueue_allocate failed to alloc vq[0]\n");
+		return;
+	}
+	vq[1] = virtqueue_allocate(VRING_SIZE);
+	if (vq[1] == NULL) {
+		printk("virtqueue_allocate failed to alloc vq[1]\n");
+		return;
+	}
+
+	vdev.role = RPMSG_MASTER;
+	vdev.vrings_num = VRING_COUNT;
+	vdev.func = &dispatch;
+	rvrings[0].io = io;
+	rvrings[0].info.vaddr = (void *)VRING_TX_ADDRESS;
+	rvrings[0].info.num_descs = VRING_SIZE;
+	rvrings[0].info.align = VRING_ALIGNMENT;
+	rvrings[0].vq = vq[0];
+
+	rvrings[1].io = io;
+	rvrings[1].info.vaddr = (void *)VRING_RX_ADDRESS;
+	rvrings[1].info.num_descs = VRING_SIZE;
+	rvrings[1].info.align = VRING_ALIGNMENT;
+	rvrings[1].vq = vq[1];
+
+	vdev.vrings_info = &rvrings[0];
+
+	/* setup rvdev */
+	rpmsg_virtio_init_shm_pool(&shpool, (void *)SHM_START_ADDR, SHM_SIZE);
+	status = rpmsg_init_vdev(&rvdev, &vdev, ns_bind_cb, io, &shpool);
+	if (status != 0) {
+		printk("rpmsg_init_vdev failed %d\n", status);
+		return;
+	}
+
+	/* Since we are using name service, we need to wait for a response
+	 * from NS setup and than we need to process it
+	 */
```

Comment:
just for you information, NS announcement is optional, you can also define a fixed source and destination address on end point creation.